### PR TITLE
Add caching to CompilerFilter to not re-compile unchanged files (rebased)

### PR DESCRIPTION
**Pull request by karyon**
this is a rebase of #346 plus a fix that was necessary to make it work for me.

please let me know whether this kind of functionality can be included and what is missing in this PR, if anything.

### CODE REVIEW

File: compressor/filters/base.py [hunk 1]
```diff
@@ -208,5 +211,26 @@ def input(self, **kwargs):
...
+
+    def __init__(self, mimetype, **kwargs):
+        self.mimetype = mimetype
+        super(CachedCompilerFilter, self).__init__(**kwargs)
+
+    def input(self, **kwargs):
+        if self.mimetype in settings.COMPRESS_CACHEABLE_PRECOMPILERS:
+            key = self.get_cache_key()
+            data = cache.get(key)
+            if data:
```
Comment (diox) [inline compressor/filters/base.py]: Might want to check for `not None` instead to avoid recompiling empty files here

Comment (diox) [inline compressor/filters/base.py]: If you do, add a test

**Comment by diox** [comment]
There are test failures, look at https://travis-ci.org/django-compressor/django-compressor/builds/78987733

Overall this looks good. I left a few comments for you to fix, but we can merge once you fix them and get tests to pass

File: compressor/filters/base.py (hunk 1 above)
Comment (karyon) [inline compressor/filters/base.py]: could you please elaborate what the test should test?

Comment (diox) [inline compressor/filters/base.py]: add a test that checks that an empty output is cached correctly (i.e. that you are not recompiling empty files)

**Comment by karyon** [comment]
could you check whether that's ok? i'd then squash most of my new commits before you merge

**Comment by diox** [comment]
Looks good to me! You can squash all your commits together.

**Comment by karyon** [comment]
done.

**Comment by diox** [comment]
Thanks!

**Comment by kivikakk** [comment]
Just what I needed! Thanks. :)
